rpcs3/Emu/RSX/rsx_cache.h

```diff
@@ -494,73 +495,99 @@ namespace rsx
 			dlg->update_msg(0, getMessage(0, 0, entry_count));
 			dlg->update_msg(1, getMessage(0, 0, entry_count));
 
-			// Setup worker threads
-			unsigned nb_threads = std::thread::hardware_concurrency();
-			std::vector<std::thread> worker_threads(nb_threads);
-
 			// Preload everything needed to compile the shaders
-			// Can probably be parallelized too, but since it's mostly reading files it's probably not worth it
 			std::vector<std::tuple<pipeline_storage_type, RSXVertexProgram, RSXFragmentProgram>> unpacked;
 			std::chrono::time_point<steady_clock> last_update;
 			u32 processed_since_last_update = 0;
-
-			for (u32 i = 0; (i < entry_count) && !Emu.IsStopped(); i++)
+						
+			if (g_cfg.video.renderer == video_renderer::vulkan)
```

Comment:
With the new code, this function has ballooned to unmanageable size. Refactor the tasks here into manageable pieces. Just add a few private methods load_shaders and compile_shaders that take an optional number of worker threads. You can pass 1 for OGL. You may keep the special handling of 1 thread for now, I'll remove it later; I know there are issues with OpenGL and context threading.

Reply:
Should be better now but OpenGL is broken since it needs setting up context before and I'm not sure how to go about it the right way.

Reply:
Hence my suggestion to keep the special handling for the thread_count == 1 case, I'll rewrite it properly later.